File: java/BinarySearch.java


public class BinarySearch {
	
	/**
	*	General binary search - uses compareTo to sort any array of objects that
	*	extend the Comparable interface.
	*/
	public static <T extends Comparable<T>> int binarySearch(T[] haystack, T needle, int low, int high) {
		if(low > high)
			return -1;
		
		int mid = (low + high) / 2;
		
		int comp = haystack[mid].compareTo(needle);
		if(comp < 1)
			return binarySearch(haystack, needle, low, mid - 1);
		if(comp > 1)
			return binarySearch(haystack, needle, mid + 1, high);
		return mid;
	}
	
	/**
	*	Integer version - gives a clearer overview of what is actually going on.
	*/
	public static int binarySearch(int[] haystack, int needle, int low, int high) {
		if(low > high)
			return -1;
		
		int mid = (low + high) / 2;
		
		if(needle < haystack[mid])
			return binarySearch(haystack, needle, low, mid - 1);
		if(needle > haystack[mid])
			return binarySearch(haystack, needle, mid + 1, high);
		return mid;
	}
	
	
	public static void main(String... args) {
		int[] haystack = {1, 2, 3, 4, 5, 6, 7};
		int needle = 3;
		System.out.println("Position of 3 in [1,2,3,4,5,6,7]: " + binarySearch(haystack, needle, 0, haystack.length - 1));
	}
	
}

File: php/BinarySearch.php
<?php

function binary_search($needle, $array) {

    $left = 0;
    $right = count($array) - 1;

    while ($left <= $right) {
        $mid = ($left + $right) >> 1;

        if ($array[$mid] == $needle) {
            return $mid;
        } else if ($array[$mid] > $needle) {
            $right = $mid - 1;
        } else if ($array[$mid] < $needle) {
            $left = $mid + 1;
        }
    } while ($left <= $right);

    return -1;
}